src/librustc/traits/fulfill.rs

```diff
@@ -341,177 +290,39 @@ impl<'a, 'gcx, 'tcx> FulfillmentContext<'tcx> {
-}
-
-impl<'a, 'b, 'gcx, 'tcx> AncestorSet<'b, 'tcx> {
-    fn new(backtrace: &Backtrace<'b, PendingPredicateObligation<'tcx>>) -> Self {
-        AncestorSet {
-            populated: false,
-            cache: FnvHashMap(),
-            backtrace: backtrace.clone(),
-        }
+    fn process_obligation(&mut self,
+                          obligation: &mut Self::Obligation)
+                          -> Result<Option<Vec<Self::Obligation>>, Self::Error>
+    {
+        process_predicate(self.selcx,
+                          obligation,
+                          self.region_obligations,
+                          self.rfc1592_obligations)
+            .map(|os| os.map(|os| os.into_iter().map(|o| PendingPredicateObligation {
+                obligation: o,
+                stalled_on: vec![]
+            }).collect()))
     }
 
-    /// Checks whether any of the ancestors in the backtrace are equal
-    /// to `predicate` (`predicate` is assumed to be fully
-    /// type-resolved).  Returns `None` if not; otherwise, returns
-    /// `Some` with the index within the backtrace.
-    fn has(&mut self,
-           infcx: &InferCtxt<'a, 'gcx, 'tcx>,
-           predicate: &ty::Predicate<'tcx>)
-           -> Option<usize> {
-        // the first time, we have to populate the cache
-        if !self.populated {
-            let backtrace = self.backtrace.clone();
-            for (index, ancestor) in backtrace.enumerate() {
-                // Ugh. This just feels ridiculously
-                // inefficient.  But we need to compare
-                // predicates without being concerned about
-                // the vagaries of type inference, so for now
-                // just ensure that they are always
-                // up-to-date. (I suppose we could just use a
-                // snapshot and check if they are unifiable?)
-                let resolved_predicate =
-                    infcx.resolve_type_vars_if_possible(
-                        &ancestor.obligation.predicate);
-
-                // Though we try to avoid it, it can happen that a
-                // cycle already exists in the predecessors. This
-                // happens if the type variables were not fully known
-                // at the time that the ancestors were pushed. We'll
-                // just ignore such cycles for now, on the premise
-                // that they will repeat themselves and we'll deal
-                // with them properly then.
-                self.cache.entry(resolved_predicate)
-                          .or_insert(index);
-            }
-            self.populated = true;
+    fn process_backedge<'c, I>(&mut self, cycle: I)
+        where I: Clone + Iterator<Item=*const Self::Obligation>,
+    {
+        if coinductive_match(self.selcx, cycle.clone()) {
+            debug!("process_child_obligations: coinductive match");
+        } else {
+            let cycle : Vec<_> = cycle.map(|c| unsafe { &*c }.obligation.clone()).collect();
```

Comment:
this unsafe block is less nice :) I thought you said you had found a workaround here?